flux-accounting DB: add `active` column

cmoussa1 opened the pull request:
#### Problem

Mentioned in #232, we need a method for "disabling" a user/bank entry in the multi-factor priority plugin's internal map when a user gets removed from a bank in the flux-accounting DB. Since entries don't get deleted in the internal map, a user can technically still run jobs under a bank they no longer belong to.

I think a potential solution to this would be to signal to the plugin when a user/bank association gets deleted in the flux-accounting DB. A `1` could represent a "deleted" user/bank row or a "deleted" bank row. Since this involves a little rework to the DB, its commands, **and** the priority plugin, I think I should break up this work into at least two separate PR's.

---

This [WIP] PR adds a new column to the `association_table` and `bank_table` in the flux-accounting DB called `deleted`, an integer column that represents when a row has been "deleted." When a user/bank row or a bank row gets "deleted," this column is updated to `1`. 

In the weighted tree library, when the data from the flux-accounting DB is loaded, if a row's `deleted` column is 1, that row is skipped and not added to the tree. 

A couple of updates to the tests that use the `delete-user` and `delete-bank` commands are also updated to accommodate this new functionality. 

A couple of new tests are also added to `t1000-print-hierarchy.t` to make sure that when a user/bank row or a bank row is "deleted," they do not show up in the output of the `flux account-shares` command (the command that prints the DB hierarchy).

cmoussa1 committed:
association_table: add 'active' column

Add a new column to the association_table: active, an integer which represents
if a user/bank association is active in the database. If the user is no longer
active, this column will show a 0.

cmoussa1 committed:
delete_user: update 'active' column

Instead of removing the user/bank association with a DELETE statement, UPDATE
the column by setting the 'active' column to 0 in the association_table.

cmoussa1 committed:
test_user_subcommands: update delete-user tests

Instead of checking for the number of rows returned by the SQLite entry,
check the value of the 'active' column to ensure that the row was disabled.

cmoussa1 committed:
t1007: change delete-user tests

Change the checks in the sharness test t1007-flux-account.t to check the value
of the 'active' column to ensure that the row was disabled.

cmoussa1 committed:
data_reader_db: only add user row if active

Check for the 'active' column in the user/bank row when adding it to the
weighted tree; if the column is 0, do not add the row.

cmoussa1 committed:
t1000: add new test for printing DB hierarchy

Add a new test which prints a flux-accounting DB hierarchy after deleting a
user to ensure that deleted users do not show up in the output of the 'flux
account-shares' command.

cmoussa1 committed:
bank_table: add 'active' column

Add a new column to the bank_table: active, an integer which represents if a
bank has been disabled in the database. If disabled, this will show a 0.

cmoussa1 committed:
delete_bank: update 'active' column

Instead of removing the bank with a DELETE statement, UPDATE the column by
setting the 'active' column to 0 in the bank_table.

cmoussa1 committed:
test_bank_subcommands: update delete-bank tests

Instead of checking for the number of rows returned by the SQLite query,
check the value of the 'active' column to ensure that the row has been disabled.

cmoussa1 committed:
t1007: update delete-bank tests

Update the tests and expected files in t1007-flux-account.t to include the
'active' column and to check the value of this column to determine if a bank
has been disabled from the flux-accounting DB.

cmoussa1 committed:
data_reader_db: only add bank row if not disabled

Check for the 'active' column in the bank row when adding it to the
weighted tree; if the column is 0, do not add the row.

cmoussa1 committed:
t1000: add new test for printing DB hierarchy

Add a new test which prints a flux-accounting DB hierarchy after deleting a
bank to ensure that deleted banks and their users do not show up in the
output of flux account-shares.

cmoussa1 committed:
t: add 'active' column to test DB's

cmoussa1 commented:
I think this is getting a little closer, so I'll take this out of [WIP] and mark it as ready for some initial review whenever convenient.

cmoussa1 committed:
add-user: check if row already exists

Before adding a user/bank row to the association_table, check if the row
already exists before adding. If it does, just UPDATE the 'active' column to
1 to re-enable the user/bank row.

cmoussa1 committed:
delete_user: update default_bank on condition

In the delete_user function, add a check to see if the bank being deleted
is a user's default bank; if it is, update the default bank in all of the
user's other rows in the association_table to be the next associated bank that
the user belongs to.

cmoussa1 committed:
t: add tests for updating default_bank on delete

Add both a unit test and a sharness test to make sure that the default bank for
a user gets updated when their default bank row gets deleted from the
association_table.

cmoussa1 commented:
Rebased to catch up after #241

chu11 commented:
an initial high level comment, would a field called `active` (or similar) be more desirable?  can a user be re-added to a bank?

chu11 commented:
> I could see using a field called active being easier to look at and keep track of, so thanks for this suggestion. I'm assuming we could use similar integer values to represent a row if it is "active" or not, 1 for active and 0 for inactive. If this makes more sense and is what you were thinking, I can change the direction of this PR to instead add an active column instead of a deleted column.

well to me "active" or perhaps "enabled" seems more binary, like user could be enabled or not, where as "deleted" sounds more permanent.  I haven't reviewed the PR yet so there could be a subtlety I'm missing.

Perhaps @grondo or @ryanday36 can give a vote as well.  It's obviously just a semantics thing and I wouldn't want to cause you too much work if it's just me.

cmoussa1 commented:
Thanks for the input @chu11 and @ryanday36, all good points. Why don't I take some time to rework this PR to instead use `active` to distinguish a user/bank row in the flux-accounting database. I can convert this back to [WIP] and open it up for review once I've made the appropriate changes.

cmoussa1 commented:
Sorry @chu11 that it took me a couple days to get back to this and to your suggestions - got sidetracked with other bug fixes in the meantime. :-) I've adjusted this PR to instead add an `active` column instead of a `deleted` column.

One thing I noticed while working on the change was that the branch name is no longer a good fit (`add.deleted.col`). I can change the name of the branch, but in doing so I _think_ it will close this PR and open a new one. If that sounds like an OK plan to you, I'll just close this PR and re-open with a better branch name.

chu11 commented on src/bindings/python/fluxacct/accounting/user_subcommands.py:
```diff
@@ -158,6 +158,24 @@ def delete_user(conn, username, bank):
+    rows = cur.fetchall()
+    default_bank = rows[0][0]
+
+    if default_bank == bank:
+        # get first bank from other potential existing rows from user
+        select_stmt = """SELECT bank FROM association_table WHERE active=1 AND username=?
+                         ORDER BY creation_time"""
+        cur.execute(select_stmt, (username,))
+        rows = cur.fetchall()
+        # update user rows to have a new default bank (the next earliest user/bank row created)
+        if len(rows) > 0:
+            new_default_bank = rows[0][0]
+            edit_user(conn, username, default_bank=new_default_bank)
```
what if the user is not a member of any other bank?  should default_bank be set to `None` or something like that?

and would test in follow on test be needed?

cmoussa1 replied:
I'm sorry, I think I am just confused by your question and am probably not answering what you are actually asking. If a user only belongs to one bank (thus, their default bank is the same as the bank they belong to), and the user gets disabled, then their default bank does not need to be updated at all. It will keep the same default bank it had before the user got disabled because the user does not belong to any other banks. Does that make sense? :P

chu11 replied:
ohh ok, that makes sense :-)  I'll let ya decide if a comment is warranted, maybe I'm just a dummy :-)

cmoussa1 replied:
Haha no you're not - I probably just didn't do a good job explaining it. I'll add a detailed comment just above explaining this case so that if/when other people look at it they won't raise similar questions.

chu11 commented on src/fairness/reader/data_reader_db.cpp:
```diff
@@ -245,33 +245,36 @@ std::shared_ptr<weighted_tree_node_t> data_reader_db_t::get_sub_banks (
-                // add association as a child of the node
-                if (add_assoc (username, shrs, usage, fshare, node) < 0) {
-                    m_err_msg += "Failed to add association\n";
+            if (active == 1) {
```
nit: doesn't have to be == 1, just `active`?

chu11 commented on src/bindings/python/fluxacct/accounting/test/test_bank_subcommands.py:
```diff
@@ -72,12 +72,13 @@ def test_04_add_subaccounts(self):
         self.assertEqual(len(rows), 1)
 
-    # removing a bank currently in the bank_table
+    # remove a bank currently in the bank_table
```
remove -> disable?

cmoussa1 commented:
Thanks for the great feedback @chu11! I went ahead and force-pushed some changes to this branch based on your suggestions:

- added a (hopefully) clear comment in `delete_user()` that talks about the behavior when a user who only belongs to one bank is disabled
- changed the check for `active == 1` to just `active` in `data_reader_db.cpp`
- changed comments that say `delete` to `disable`

cmoussa1 committed:
t1017: add 'active' to expected columns list

chu11 reviewed (approved):
just the minor nits I mentioned above, I think its good to go

cmoussa1 commented:
Just pushed up a change to address @chu11's good suggestions above and added some additional comments to hopefully help clarify the expected behavior when a user gets disabled while only belonging to one bank. I'll set MWP shortly
